### PR TITLE
feat(hexo-next): compatable with client@v2

### PR DESCRIPTION
**Pull request by Mister-Hope**

### CODE REVIEW

**Comment by leirock** [comment]
已测试，有报错。情况如下：

【尝试1】使用该 PR 版本，终端会报错如下。

```
Unhandled rejection Template render error: (/Users/leirock/Git/blog/node_modules/hexo-theme-next/layout/page.njk)
  Template render error: (/Users/leirock/Git/blog/node_modules/hexo-theme-next/layout/page.njk) [Line 36, Column 25]
  Template render error: (/Users/leirock/Git/blog/node_modules/hexo-theme-next/layout/page.njk) [Line 49, Column 17]

---省略部分类似报错内容----

  Template render error: (/Users/leirock/Git/blog/node_modules/hexo-theme-next/layout/inject/bodyEnd/waline.njk) [Line 6, Column 65]
  TypeError: Cannot destructure property 'url' of 'this.theme.vendors[name]' as it is undefined.
    at Object._prettifyError (/Users/leirock/Git/blog/node_modules/nunjucks/src/lib.js:36:11)
    at /Users/leirock/Git/blog/node_modules/nunjucks/src/environment.js:563:19
    at Template.root [as rootRenderFunc] (eval at _compile (/Users/leirock/Git/blog/node_modules/nunjucks/src/environment.js:633:18), <anonymous>:45:3)
    at Template.render (/Users/leirock/Git/blog/node_modules/nunjucks/src/environment.js:552:10)
    at /Users/leirock/Git/blog/node_modules/hexo/lib/plugins/renderer/nunjucks.js:67:37
    at _View._compiled (/Users/leirock/Git/blog/node_modules/hexo/lib/theme/view.js:136:50)
    at _View.render (/Users/leirock/Git/blog/node_modules/hexo/lib/theme/view.js:39:17)
    at /Users/leirock/Git/blog/node_modules/hexo/lib/hexo/index.js:64:21
    at tryCatcher (/Users/leirock/Git/blog/node_modules/bluebird/js/release/util.js:16:23)
    at /Users/leirock/Git/blog/node_modules/bluebird/js/release/method.js:15:34
    at RouteStream._read (/Users/leirock/Git/blog/node_modules/hexo/lib/hexo/router.js:47:5)
    at RouteStream.Readable.read (node:internal/streams/readable:487:10)
    at resume_ (node:internal/streams/readable:987:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21)
```

【尝试2】据此，修改 `waline.njk` ，将 `theme.vendors.` 和 `next_vendors()` 相关的全部删去，直接定义 `waline_css` 和 `waline_uri`

```diff 
- {%- set waline_css_uri = theme.vendors.waline_css or next_vendors('//cdn.jsdelivr.net/npm/@waline/client@v2/dist/waline.css') %}
+ {%- set waline_css_uri = '//cdn.jsdelivr.net/npm/@waline/client@v2/dist/waline.css' %}

- {%- set waline_uri = theme.vendors.waline or next_vendors('//cdn.jsdelivr.net/npm/@waline/client@v2') %}
+ {%- set waline_uri = '//cdn.jsdelivr.net/npm/@waline/client@v2' %}
```

这种情况下，终端不报错，浏览器报错如下：

```
[Microsoft Edge]
Uncaught SyntaxError: missing ) after argument list

[macOS Safari]
SyntaxError: Unexpected token ';'. Expected ')' to end an argument list.
```

【尝试3】然后继续修改 `waline.njk`，并在 `default.yaml` 设置好默认的 `waline.cssUrl = //cdn.jsdelivr.net/npm/@waline/client@v2/dist/waline.css` ：

```diff
{{ next_data('waline', {
  lang: page.lang | default('zh-CN', true)
}, config.waline, {
  comment: true,
+  jsUrl: config.waline.jsUrl | default('//cdn.jsdelivr.net/npm/@waline/client@v2/dist/waline.js', true),
  path: url_for(page.path) | replace(r/\/index\.html$/, '/')
}) }}

- {%- set waline_css_uri = theme.vendors.waline_css or next_vendors('//cdn.jsdelivr.net/npm/@waline/client@v2/dist/waline.css') %}
- <link rel="stylesheet" href="{{ waline_css_uri }}">
+ <link rel="stylesheet" href="{{ config.waline.cssUrl }}">
<script>
document.addEventListener('page:loaded', () => {
  NexT.utils.loadComments(CONFIG.waline.el).then(() =>
-    NexT.utils.getScript({{ waline_uri }}, { condition: window.Waline })
+    NexT.utils.getScript(CONFIG.waline.jsUrl, { condition: window.Waline })
  ).then(() => Waline.init(CONFIG.waline));
});
</script>
```

此时，Edge 浏览器可以正常加载评论，但是 **Safari 浏览器无法加载评论** ，并报错如下：
截屏2022-04-26 02 20 00

此外，此种情形下，均 **无法在主页的文章列表中加载评论统计的数字** （但是 Edge 浏览器可以在文章页的标题栏下加载评论统计数字）。

File: packages/hexo-next/index.js [hunk 1]
```diff
@@ -74,25 +69,27 @@ hexo.extend.filter.register('theme_inject', (injects) => {
...
+      'waline_pageview',
       `
-    <span id="{{ url_for(post.path) }}" class="post-meta-item leancloud_visitors" data-flag-title="{{ post.title }}" title="{{ __('post.views') }}">
+    <span id="{{ url_for(post.path) }}" class="post-meta-item" title="{{ __('post.views') }}">
       <span class="post-meta-item-icon">
         <i class="far fa-eye"></i>
       </span>
       <span class="post-meta-item-text">{{ __('post.views') + __('symbol.colon') }}</span>
-      <span class="leancloud-visitors-count"></span>
+      <span class="waline-pagwview-count"></span>
```
Comment (Mister-Hope) [inline packages/hexo-next/index.js:92]: Ah hhhhh

Comment (Mister-Hope) [reply packages/hexo-next/index.js]: A silly typo.

**Comment by Mister-Hope** [comment]
@leirock  would you please check again?

**Comment by leirock** [comment]
> @leirock Could you please check again?

The `libUrl` should be `https://cdn.jsdelivr.net/npm/@waline/client@v2/dist/waline.js`. Otherwise, Waline will not be loaded.

## My modifications

- **waline.njk**


```diff
{{ next_data('waline', {
  lang: page.lang | default('zh-CN', true)
}, config.waline, {
  comment: true,
- libUrl: config.waline.libUrl | default('https://cdn.jsdelivr.net/npm/@waline/client@v2', true),
+ libUrl: config.waline.libUrl | default('https://cdn.jsdelivr.net/npm/@waline/client@v2/dist/waline.js', true),
- cssUrl: config.waline.libUrl | default('https://cdn.jsdelivr.net/npm/@waline/client@v2/dist/waline.css', true),
  path: url_for(page.path) | replace(r/\/index\.html$/, '/')
}) }}
<link rel="stylesheet" href="{{ config.waline.cssUrl }}">
<script>
document.addEventListener('page:loaded', () => {
  NexT.utils.loadComments(CONFIG.waline.el).then(() =>
    NexT.utils.getScript(CONFIG.waline.libUrl, { condition: window.Waline })
  ).then(() => Waline.init(CONFIG.waline));
});
</script>
``` 
Remarks: `cssUrl` is not required for Waline configuation, and the default value of `cssUrl` can be defined in `default.yaml`

- **default.yaml**

```diff
+ cssUrl: https://cdn.jsdelivr.net/npm/@waline/client@v2/dist/waline.css
```

## Remained issues after modifications

- Numbers of **comments** and **pageview** cannot be loaded on the index page (home page)  in post meta. But they can be loaded normally on the post page in post meta.
- `copyright: false` is not valid.

**Comment by leirock** [comment]
> could you please give some info about home page structure? Is the selector valid or not?

@Mister-Hope The selector is valid in the post page but it seems invalid in the home page. Anyway, I have sent my blog source to your email shown in your GitHub profile.

**Comment by Mister-Hope** [comment]
Already published to main. Feel free to open issues if you can not get it work with main branch code

**Pull request closed without merging**